### ISSUE TITLE
lagging and slowness when using large photos

### ISSUE THREAD

**Issue by xperator** [Norserium/vue-advanced-cropper#134]
Hey,
First of all, great library. I'm loving it ❤ and actually using it in my first big project which depends on working with lots of images.

I have a huge problem with this library and that is when I use a large photo ( like anything over 2000px or over 2Mb of file size ) the editor gets super laggy and everything becomes super slow. The editor just becomes very unpleasant to use.

I've checked the docs multiple times but couldn't find any recipe or notes on how to work with large images (or maybe I miss it?)

Am I missing something? Am I not supposed to use large photos with this library? Maybe I should downsize the photo before load it with the library?

Any help would be appreciated. Thanks in advance

**Comment by Norserium** [comment]
Hello, @xperator! 

How do you upload that large photos? 

It's recommended to use `URL.createObjectURL` to receive the direct link to the file, instead of using `reader.readAsDataURL`. [The example](https://norserium.github.io/vue-advanced-cropper/guides/recipes.html#load-image-from-a-disc) is in the documentation.

**Comment by xperator** [comment]
@Norserium Thanks for reaching out quickly!

Well I have a "Upload File" button like so:
```vue
<b-upload
    v-model="image_file"
    @input="loadImage"
    accept="image/*"
    drag-drop>
    Upload File
</b-upload>
```
and I setup cropper like this:
```vue
<cropper
    ref="cropper"
    class="cropper"
    :src="image"
    :auto-zoom="true"
    minWidth="1440"
    minHeight="1080"
    :stencil-props="{
        aspectRatio: 4/3
    }"
    image-restriction="stencil"/>
```
And this is my `loadImage` function:
```js
loadImage() {
    var reader = new FileReader();
    reader.onload = (e) => {
        this.image = e.target.result;
    };
    reader.readAsDataURL(this.image_file);
},
```

The method you just mentioned, I thought that is for when using a direct web image url instead of user uploaded image?

**Comment by Norserium** [comment]
No, it's not. There is real-time example, give it a try with your large images.

**Comment by xperator** [comment]
@Norserium Yeah I just tried it lots of images and it works flawlessly 😄 ( even with big 15Megabytes photos)
I may be totally off but have you changed the docs recently? I remember the "Load image from a disc" recipe was a bit different when I checked it a while ago.

Anyway I got to figure out how to use the `URL.createObjectURL` in my `loadImage` function.
Thanks for the help ! Great stuff 🚀👏

**Comment by Norserium** [comment]
>  I remember the "Load image from a disc" recipe was a bit different when I checked it a while ago.

Yes, you are right. It was updated 15 days ago.

> Update: I changed it to this

Why do you need this fragment?
```js
    var reader = new FileReader();
    reader.onload = (e) => {
        this.image = blob;
    };
    reader.readAsDataURL(this.image_file);
```

It's useless. I added the reading file content only to parse the actual file format. If you don't need it, you can short your code:
```js
loadImage() {
    this.image = URL.createObjectURL(this.image_file);
},
```

**Comment by xperator** [comment]
Yeah probably didn't need that. I'm respectfully using your code now. Thanks again.

My "issue" is solved for this specific topic so you can close it if you want.
I have an off topic question if you don't mind and have some spare time 😇
I need my users to upload 2 different aspect ratio of the same photo. Like let's say one is a landscape ( 1440x1080 ) and the other has to be portrait ( 500x565 ). My guess is that I need to have 2 different editors and have the second one just load the same photo from the first one and have it use different configuration, right?

**Comment by Norserium** [comment]
> My guess is that I need to have 2 different editors and have the second one just load the same photo from the first one and have it use different configuration, right?

There is no a definite answer. Two different editors looks like the simplest way, I don't see any crucial disadvantages of this method. As an alternative you can try to change `aspectRatio` dynamically, it will work also.

**Comment by xperator** [comment]
Hey,
I'm currently having same issue with a few specific images. When I load them in the [example](https://norserium.github.io/vue-advanced-cropper/guides/recipes.html#load-image-from-a-disc) they work fine.
I don't know what is causing the issue. I tried removing all extra css and all extra cropper configuation. The code for loading the image is the same as above.

**Comment by Norserium** [comment]
@xperator, could you provide that specific images? How can I reproduce this error?

**Comment by xperator** [comment]
> @xperator, could you provide that specific images? How can I reproduce this error?

Well I edited my comment earlier, and this seems to be caused by uBlock Origin browser extension. When I disable it and refresh the page, the cropper will work fine with any image.
I'm trying to figure out on how to fix this without disabling uBlock extension

**Comment by Norserium** [comment]
@xperator, but why it works fine on my example site? What's the difference?

**Comment by xperator** [comment]
> @xperator, but why it works fine on my example site? What's the difference?

That's what I dont understand either. It's somehow interfering with my local dev site but your public online example site is working just as fine. Maybe either something on your site makes it green so to speak, or something on my part is making uBlock getting picky about it.

**Comment by Norserium** [comment]
Do you use `https` protocol locally?

**Comment by xperator** [comment]
> Do you use `https` protocol locally?

No. I never tried to do ssl locally cause I thought it didn't matter.

**Comment by Norserium** [comment]
I'm not sure that it's the reason of that lags, but it theoretically may be.

**Comment by xperator** [comment]
Now disabling uBlock doesn't seem to be working... And it's not working with your online example site either.. I'm trying to figure out what the heck is going atm... 
[Here](http://interfacelift.com/wallpaper/D2921381/03935_empireofthealps_3840x2160.jpg) ( I downloaded the 3840x2160 one ) is a link to one of the images if you wanna try it yourself but i'm not sure if it will be reproduced on your side.

**Comment by Norserium** [comment]
@xperator, actually, I figured out the reason. It's internal. You can currently disable `check-orientation` and it will solve your problem. At the same time I will try to deal with it.

**Comment by xperator** [comment]
> @xperator, actually, I figured out the reason. It's the internal issue. You can currently disable `check-orientation` and it will solve your problem. At the same time I will try to figure out how to solve it.

That's good news! 😄👌
Was the image I shared any helpful at least or you figured out the reason by other means?

I was just scratching my head for hours. No wonder this issue occurring was felt random. Sometimes it wouldn't happen, and sometimes it would come back no matter what you changed or tried.

**Comment by Norserium** [comment]
@xperator, yes it was helpful. It gave a clue. I assume that you encounter this problem with images that has `EXIF`.

**Comment by xperator** [comment]
If you don't mind, I also encountered another issue last night. I wasn't sure if I should open up a new topic for this or just comment it here.

The issue is that when I load an image, the width of the cropper exceed all ancestor element widths and it causes layout issues and horizontal scrolling. I'm talking about the dynamically calculated widths that are added in-line to elements like `vue-advanced-cropper__stretcher` and `vue-advanced-cropper__boundaries`.
I don't know if something is causing the width to be calculated wrong or it's something I have on my css part that is causing this.

I have tried lots of different things but nothing worked except when I remove the `display: flex` from the cropper parent, the width will be normal and never exceeds it's parents. This is kind of weird because the parent that has flex display is not even a direct parent of the cropper. There is like 4-5 level difference.

Any ideas or clues? Thanks again.

**Comment by Norserium** [comment]
@xperator, could you create the codesandbox that reproduces this issue?

**Comment by xperator** [comment]
> @xperator, could you create the codesandbox that reproduces this issue?

Give me some time maybe I can put something together. I currently ran into a CORS issue with my app and it doesn't let me login into the page where I had the issue ( Some annoying Nginx/Laravel config issue I believe )

**Comment by Norserium** [comment]
@xperator, try to login in Firefox.

**Comment by Norserium** [comment]
@xperator, try to update to `1.5.1`. It should fix the original issue.

**Comment by xperator** [comment]
> @xperator, try to login in Firefox (in the case, if other requests work well)

No it's not working there either. It's funny cause it was working fine in Chrome at first and it started throwing CORS errors in FF and now Chrome is doing the same thing as well.
I've been working on a Nuxt ( SSR ) + Laravel ( as API backend ) project for few months, but I still haven't got a stable Nginx configuation yet. There are not many tutorials available so I'm just gluing random things together and hoping for the best.

**Comment by Norserium** [comment]
> I have tried lots of different things but nothing worked except when I remove the display: flex from the cropper parent, the width will be normal and never exceeds it's parents. 

Did you try to set `min-width: 0` to the cropper?

**Comment by Norserium** [comment]
> I've been working on a Nuxt ( SSR ) + Laravel ( as API backend ) project for few months, but I still haven't got a stable Nginx configuation yet. There are not many tutorials available so I'm just gluing random things together and hoping for the best.

What's the error? What's your configuartion?

**Comment by xperator** [comment]
> Did you try to set `min-width: 0` to the cropper?

No but I will.

> What's the error? What's your configuartion? Why do you use CORS anyway?

The basic idea is to have Nginx to act as a reverse proxy and have it `proxy_pass` all requests to Nuxt at `localhost:3000` except for the `/laravel` path where it should `proxy_pass` that one to laravel side so client can talk to laravel api endpoint.
Because I couldn't figure out a way to have all of these inside a single `server` block, I ended up separating them and have one server block for Nuxt and one for Laravel like so:

This is my current Nginx config:

```nginx
map $sent_http_content_type $expires {
    "text/html"                 epoch;
    "text/html; charset=utf-8"  epoch;
    default                     off;
}

server {
    listen       80;
    server_name  example.test;

    location /laravel/ {
        proxy_pass http://api.example.test/;
    }

    location /photos {
        root /home/myuser/code/example-laravel/storage/app/;
    }

    location / {
        expires $expires;

        proxy_redirect                      off;
        proxy_set_header Host               $host;
        proxy_set_header X-Real-IP          $remote_addr;
        proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto  $scheme;
        proxy_read_timeout          1m;
        proxy_connect_timeout       1m;
        proxy_pass   http://localhist:3000;
    }

    location ~ /\.ht {
        deny  all;
    }
}

server {
    listen 80;
    server_name api.example.test;
    root /home/myuser/code/example-laravel/public;

    add_header X-Frame-Options "SAMEORIGIN";
    add_header X-XSS-Protection "1; mode=block";
    add_header X-Content-Type-Options "nosniff";

    index index.php;

    charset utf-8;

    allow       127.0.0.1;
    deny        all;

    location / {
        try_files $uri $uri/ /index.php?$query_string;
    }

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt  { access_log off; log_not_found off; }

    error_page 404 /index.php;

    location ~ \.php$ {
        fastcgi_pass unix:/var/run/php/php8.0-fpm.sock;
        fastcgi_param SCRIPT_FILENAME $realpath_root$fastcgi_script_name;
        include fastcgi_params;
    }

    location ~ /\.(?!well-known).* {
        deny all;
    }
}
```

**Comment by xperator** [comment]
I honestly can't tell if I need CORS or not..? From what I read online I was told to have it configured to work...?

**Comment by xperator** [comment]
And the error is :
```js
Access to XMLHttpRequest at 'http://api.example.test/' (redirected from 'http://example.test/laravel/login') from origin 'http://example.test' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Comment by Norserium** [comment]
> I honestly can't tell if I need CORS or not..? 

You placed your api on the subdomain, so it's different host and CORS is needed.

In this case I don't understand why do you need this fragment:
```
    location /laravel/ {
        proxy_pass http://api.example.test/;
    }
```

**Comment by Norserium** [comment]
Try to replace this fragment:
```
    location /laravel/ {
        proxy_pass http://api.example.test/;
    }
```

By this fragment:
```
    location /laravel/ {
        proxy_pass http://api.example.test/;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_ssl_session_reuse off;
        proxy_set_header Host $http_host;
        proxy_redirect off;
    }
```

**Comment by xperator** [comment]
> Try to replace this fragment:

Thanks for helping out you're really being nice 😇 ! Well it's acting up atm and `http://example.test/laravel/` is not pointing to laravel endpoint.

**Comment by Norserium** [comment]
I.e. `http://example.test/laravel/` returns the nuxt application in the browser?

**Comment by xperator** [comment]
> I.e. `http://example.test/laravel/` returns the nuxt application in the browser?

yes. I read I have to remove trailing slash on `proxy_pass http://api.example.test/;` but even so, it throws a :
```
400 Bad Request
Request Header Or Cookie Too Large
```

**Comment by Norserium** [comment]
Could you provide the request headers?

**Comment by xperator** [comment]
> Could you provide the request headers?

General:
```
Request URL: http://example.test/laravel/
Request Method: GET
Status Code: 400 Bad Request
Remote Address: 192.168.1.200:80
Referrer Policy: strict-origin-when-cross-origin
```

Request Headers:
```
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,fa;q=0.8
Cache-Control: no-cache
Connection: keep-alive
Host: example.test
Pragma: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36
```

**Comment by Norserium** [comment]
Perhaps, it's loop redirect. Look in `/var/log/nginx/error.log` the last request.

**Comment by Norserium** [comment]
Could you send the last version of config? For example, this line looks suspicious:
```
proxy_pass   http://localhist:3000;
```

**Comment by Norserium** [comment]
Also, fix the mentioned above fragment:
```
    location /laravel/ {
        proxy_pass http://api.example.test/;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_ssl_session_reuse off;
        proxy_redirect off;
    }
```

**Comment by xperator** [comment]
@Norserium It is working, for now! 😅 Thanks a lot! I was trying tons of different nginx config while you were helping out.
So apparently I had the `proxy_set_header Host $http_host;` added which was one of the issues so I removed it.

> Could you send the last version of config? For example, this line looks suspicious:
> 
> ```
> proxy_pass   http://localhist:3000;
> ```

Yeah I had that modified before posting it here. I usually change names and ip's and all the environment specific configuration to avoid confusion so we can focus on real issues.

> Also, replace the mentioned above fragment by this:

Totally working now 👍
Not to mention I also had to clear all cookies and website data + a system reboot.

**Comment by Norserium** [comment]
> So apparently I had the proxy_set_header Host $http_host; added which was one of the issues so I removed it.

Changing the `Host` cause that proxy request fall into the same server block, but you have backend in the another server block.

> I remember a few while ago I read something about this min-width: 0 to fix some of flexbox issues and now that you mentioned it I can recall ok_hand

By default, `min-width` of flex child is `auto`, that means, that it will preserve its original size. Even if it larger than container.

> Thanks buddy you really helped me tackle few on and off topic issues innocentclaprocket

You are welcome!
